src/realm/realm.h

```diff
@@ -0,0 +1,1612 @@
+    realm_value_type_e type;
+} realm_value_t;
+
+
+/* Error types */
+typedef struct realm_async_error realm_async_error_t;
+typedef enum realm_errno {
+    RLM_ERR_NONE = 0,
+    RLM_ERR_UNKNOWN,
+    RLM_ERR_OTHER_EXCEPTION,
+    RLM_ERR_OUT_OF_MEMORY,
+    RLM_ERR_NOT_CLONABLE,
+
+    RLM_ERR_INVALIDATED_OBJECT,
+    RLM_ERR_INVALID_PROPERTY,
+    RLM_ERR_MISSING_PROPERTY_VALUE,
+    RLM_ERR_PROPERTY_TYPE_MISMATCH,
+    RLM_ERR_MISSING_PRIMARY_KEY,
+    RLM_ERR_WRONG_PRIMARY_KEY_TYPE,
+    RLM_ERR_MODIFY_PRIMARY_KEY,
+    RLM_ERR_READ_ONLY_PROPERTY,
+    RLM_ERR_PROPERTY_NOT_NULLABLE,
+    RLM_ERR_INVALID_ARGUMENT,
+
+    RLM_ERR_LOGIC,
+    RLM_ERR_NO_SUCH_TABLE,
+    RLM_ERR_TABLE_NAME_IN_USE,
+    RLM_ERR_CROSS_TABLE_LINK_TARGET,
+    RLM_ERR_DESCRIPTOR_MISMATCH,
+    RLM_ERR_UNSUPPORTED_FILE_FORMAT_VERSION,
+    RLM_ERR_MULTIPLE_SYNC_AGENTS,
+    RLM_ERR_ADDRESS_SPACE_EXHAUSTED,
+    RLM_ERR_MAXIMUM_FILE_SIZE_EXCEEDED,
+    RLM_ERR_OUT_OF_DISK_SPACE,
+    RLM_ERR_KEY_NOT_FOUND,
+    RLM_ERR_COLUMN_NOT_FOUND,
+    RLM_ERR_COLUMN_ALREADY_EXISTS,
+    RLM_ERR_KEY_ALREADY_USED,
+    RLM_ERR_SERIALIZATION_ERROR,
+    RLM_ERR_INVALID_PATH_ERROR,
+    RLM_ERR_DUPLICATE_PRIMARY_KEY_VALUE,
+
+    RLM_ERR_INDEX_OUT_OF_BOUNDS,
+
+    RLM_ERR_INVALID_QUERY_STRING,
+    RLM_ERR_INVALID_QUERY,
+    // ...
+} realm_errno_e;
+
+typedef enum realm_logic_error_kind {
+    RLM_LOGIC_ERR_NONE = 0,
+    RLM_LOGIC_ERR_STRING_TOO_BIG,
+    // ...
+} realm_logic_error_kind_e;
+
+typedef struct realm_error {
+    realm_errno_e error;
+    realm_string_t message;
+    union {
+        int code;
+        realm_logic_error_kind_e logic_error_kind;
+    } kind;
+} realm_error_t;
+
+/* Schema types */
+
+typedef enum realm_column_attr {
+    // Values matching `realm::ColumnAttr`.
+    RLM_COLUMN_ATTR_NONE = 0,
+    RLM_COLUMN_ATTR_INDEXED = 1,
+    RLM_COLUMN_ATTR_UNIQUE = 2,
+    RLM_COLUMN_ATTR_RESERVED = 4,
+    RLM_COLUMN_ATTR_STRONG_LINKS = 8,
+    RLM_COLUMN_ATTR_NULLABLE = 16,
+    RLM_COLUMN_ATTR_LIST = 32,
+    RLM_COLUMN_ATTR_DICTIONARY = 64,
+    RLM_COLUMN_ATTR_COLLECTION = 64 + 32,
+} realm_column_attr_e;
+
+typedef enum realm_property_type {
+    // Values matching `realm::ColumnType`.
+    RLM_PROPERTY_TYPE_INT = 0,
+    RLM_PROPERTY_TYPE_BOOL = 1,
+    RLM_PROPERTY_TYPE_STRING = 2,
+    RLM_PROPERTY_TYPE_BINARY = 4,
+    RLM_PROPERTY_TYPE_MIXED = 6,
+    RLM_PROPERTY_TYPE_TIMESTAMP = 8,
+    RLM_PROPERTY_TYPE_FLOAT = 9,
+    RLM_PROPERTY_TYPE_DOUBLE = 10,
+    RLM_PROPERTY_TYPE_DECIMAL128 = 11,
+    RLM_PROPERTY_TYPE_OBJECT = 12,
+    RLM_PROPERTY_TYPE_LINKING_OBJECTS = 14,
+    RLM_PROPERTY_TYPE_OBJECT_ID = 15,
+    RLM_PROPERTY_TYPE_UUID = 17,
+} realm_property_type_e;
+
+typedef enum realm_collection_type {
+    RLM_COLLECTION_TYPE_NONE = 0,
+    RLM_COLLECTION_TYPE_LIST = 1,
+    RLM_COLLECTION_TYPE_SET = 2,
+    RLM_COLLECTION_TYPE_DICTIONARY = 4,
+} realm_collection_type_e;
+
+typedef struct realm_property_info {
+    realm_string_t name;
```

Comment:
That would require that it is zero-terminated, which is not the case if it comes from a `StringData` in Core.

Reply:
The strings returned from Core are zero-terminated.

Reply:
`StringData` makes no such guarantee - it is not generally legal to ignore the `size()`, even if most APIs happen to return zero-terminated strings. I think it would be unfortunate to commit to returning zero-terminated strings everywhere.

Reply:
Anyway, the fields in question are never assigned values coming directly from StringData. They are assigned values coming from std::string where you can just use `c_str()` to get the `const char*` value. I tried to actually make the change I suggest here, and everything just becomes very much nicer. So far, none of the arguments have convinced me that this is not a good idea.